pkg/restic/exec_commands.go

```diff
@@ -53,3 +70,109 @@ func GetSnapshotID(repoIdentifier, passwordFile string, tags map[string]string,
 
 	return snapshots[0].ShortID, nil
 }
+
+// RunBackup runs a `restic backup` command and watches the output to provide
+// progress updates to the caller.
+func RunBackup(backupCmd *Command, log logrus.FieldLogger, updateFunc func(velerov1api.PodVolumeOperationProgress)) (string, string, error) {
+	// buffers for copying command stdout/err output into
+	stdoutBuf := new(bytes.Buffer)
+	stderrBuf := new(bytes.Buffer)
+
+	// create a channel to signal when to end the goroutine scanning for progress
+	// updates
+	quit := make(chan struct{})
+	var wg sync.WaitGroup
+
+	cmd := backupCmd.Cmd()
+	cmd.Stdout = stdoutBuf
+	cmd.Stderr = stderrBuf
+
+	cmd.Start()
+
+	wg.Add(1)
+	go func() {
+		ticker := time.NewTicker(backupProgressCheckInterval)
+		for {
+			select {
+			case <-ticker.C:
+				lastLine := getLastLine(stdoutBuf.Bytes())
+				stat, err := decodeBackupStatusLine(lastLine)
+				if err != nil {
+					log.WithError(err).Errorf("error getting restic backup progress")
+				}
+
+				// if the line contains a non-empty bytes_done field, we can update the
+				// caller with the progress
+				if stat.BytesDone != 0 {
+					updateFunc(velerov1api.PodVolumeOperationProgress{
+						TotalBytes: stat.TotalBytes,
+						BytesDone:  stat.BytesDone,
+					})
+				}
+			case <-quit:
+				ticker.Stop()
+				wg.Done()
+				return
+			}
+		}
+	}()
+
+	cmd.Wait()
+	close(quit)
+	wg.Wait()
```

Comment:
I don't think we need this wait group, do we? I think the "quit" logic could be simplified - after L120, call `ticker.Stop()` (you'll need to move `ticker`'s declaration to outside the goroutine), then `close(quit)`. The `select` case that receives from `quit` can just return.

Reply:
Yeah, it's mostly left over from the additional goroutines we had for copying stdout/stderr. One issue is that if we're in the middle of the ticker.C case and it takes some time to complete, then we might call `updateFunc` in the goroutine with `BytesDone < TotalBytes` after our call to it at the end of this function. If we wait for the goroutine to finish, we know this will definitely not happen.

That said, the stuff in the `ticker.C` case should be pretty fast for backup, for restore I think we'll need the waitgroup since calculating the size of the volume can take an indeterminate amount of time.

Reply:
OK, that's a good point.  I guess the other way you could do this is to send on the `quit` channel rather than closing it, since that would block until it's received from, but I'm fine with this as-is.

Reply:
@skriss ah good point, I didn't think of that. I like the idea of removing the unnecessary waitgroup, so will update this to do that - thanks!